### PR TITLE
[TT-12775] Request size limit breaks GET and DELETE requests

### PR DESCRIPTION
**Pull request by buraksezer**
### **User description**
<details open>
  <summary><a href="https://tyktech.atlassian.net/browse/TT-12775" title="TT-12775" target="_blank">TT-12775</a></summary>
  <br />
  <table>
    <tr>
      <th>Summary</th>
      <td>Request size limit breaks GET and DELETE requests</td>
    </tr>
    <tr>
      <th>Type</th>
      <td>
        Bug
        Bug
      </td>
    </tr>
    <tr>
      <th>Status</th>
      <td>In Dev</td>
    </tr>
    <tr>
      <th>Points</th>
      <td>N/A</td>
    </tr>
    <tr>
      <th>Labels</th>
      <td><a href="https://tyktech.atlassian.net/issues?jql=project%20%3D%20TT%20AND%20labels%20%3D%20'24Bugsmash%20ORDER%20BY%20created%20DESC" title="'24Bugsmash">'24Bugsmash</a>, <a href="https://tyktech.atlassian.net/issues?jql=project%20%3D%20TT%20AND%20labels%20%3D%20customer_bug%20ORDER%20BY%20created%20DESC" title="customer_bug">customer_bug</a>, <a href="https://tyktech.atlassian.net/issues?jql=project%20%3D%20TT%20AND%20labels%20%3D%20jira_escalated%20ORDER%20BY%20created%20DESC" title="jira_escalated">jira_escalated</a></td>
    </tr>
  </table>
</details>

---

PR for https://tyktech.atlassian.net/browse/TT-12775

This PR changes behavior to set `statedCL` to zero and continue processing the request, instead of returning an error with HTTP 411. 

The issue states the following:

> The bug aside, a proposed consideration for refinement of this feature; a potential memory optimisation point -
The current implementation reads the entire request and then checks if the size limit is exceeded.
Proposal: Read requests only up to the point of the set limit.

The implementation reads the request size from `http.Request.ContentLength`, it doesn't read the whole request body to count the bytes.

___

### **PR Type**
Bug fix, Tests

___

### **Description**
- Fixed a bug where requests with missing `Content-Length` headers would break by setting `statedCL` to "0" instead of returning an error.
- Added comprehensive tests to verify that requests without `Content-Length` headers are processed correctly across various HTTP methods.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Bug fix</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>mw_request_size_limit.go</strong><dd><code>Adjust request size limit handling for missing Content-Length header</code></dd></summary>
<hr>

gateway/mw_request_size_limit.go

<li>Modified the behavior to set <code>statedCL</code> to "0" when the <code>Content-Length</code> <br>header is absent.<br> <li> Removed the error response for missing <code>Content-Length</code> header.<br>

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6734/files#diff-601cc6b053789b90f3c9fbf42b251e8e560e547981f0d17089893db743ef5b9e">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>mw_request_size_limit_test.go</strong><dd><code>Add tests for handling requests without Content-Length header</code></dd></summary>
<hr>

gateway/mw_request_size_limit_test.go

<li>Added a new test case to ensure requests do not break when <br><code>Content-Length</code> is absent.<br> <li> Tested multiple HTTP methods for this scenario.<br>

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6734/files#diff-107317fefc06776e7acf5e35daac311b025a92c6721432272dbd7c7dcdd854f8">+19/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

**Comment by buger** [comment]
<p>I'm a bot and I 👍 this PR title. 🤖</p>

**Review by kofoworola — APPROVED** [review]

**Review by titpetric — CHANGES_REQUESTED** [review]
As the bot says:

> The change from returning an error when 'Content-Length' header is missing to setting it to "0" might introduce unexpected behavior in downstream processing where actual content length is required.

As for the HTTP methods spec:

> HTTP request bodies are theoretically allowed for all methods except TRACE, however they are not commonly used except in PUT, POST and PATCH. Because of this, they may not be supported properly by some client frameworks, and you should not allow request bodies for GET, DELETE, TRACE, OPTIONS and HEAD methods.

I think the correct fix here is to ignore/skip content-length checks for the methods listed here. Content-length handling in respect to request size limit middleware if enabled, should *require content length* and respond with that 411 status code. Alternatively, wouldn't `r.Body` be nil?

**Comment by buraksezer** [comment]
> As the bot says:
> 
> > The change from returning an error when 'Content-Length' header is missing to setting it to "0" might introduce unexpected behavior in downstream processing where actual content length is required.
> 
> As for the HTTP methods spec:
> 
> > HTTP request bodies are theoretically allowed for all methods except TRACE, however they are not commonly used except in PUT, POST and PATCH. Because of this, they may not be supported properly by some client frameworks, and you should not allow request bodies for GET, DELETE, TRACE, OPTIONS and HEAD methods.
> 
> I think the correct fix here is to ignore/skip content-length checks for the methods listed here. Content-length handling in respect to request size limit middleware if enabled, should _require content length_ and respond with that 411 status code. Alternatively, wouldn't `r.Body` be nil?

Request size limit middleware skips GET, DELETE, TRACE, OPTIONS and HEAD requests. I also bring back the original error message. `Content-Length` is required for POST, PUT and PATCH methods. If it doesn't exist, the middleware returns `Content length is required for this request` with HTTP 411 code. 

> Alternatively, wouldn't `r.Body` be nil?

I checked this, `r.Body` is always initialized.

**Review by pvormste — APPROVED** [review]

**Comment by buraksezer** [comment]
Hey @titpetric, could you please review the changes again?

File: gateway/mw_request_size_limit_test.go [hunk 1]
```diff
@@ -51,4 +51,13 @@ func TestRequestSizeLimit(t *testing.T) {
 			}...)
 		})
 	})
+
+	t.Run("should not break the request, if the method is skipped", func(t *testing.T) {
+		// GET, DELETE, TRACE, OPTIONS and HEAD
```
Comment (titpetric) [inline gateway/mw_request_size_limit_test.go:56]: Add a negative test case, POST, PATCH

**Review by titpetric — APPROVED** [review]

**Pull request merged**